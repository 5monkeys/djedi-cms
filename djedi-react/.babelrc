{
  "env": {
    "development": {
      "presets": ["next/babel"],
      "plugins": ["./babel-plugin"]
    },
    "production": {
      "presets": ["next/babel"],
      "plugins": ["./babel-plugin"]
    },
    "test": {
      "presets": [
        [
          "@babel/env",
          { "loose": true, "exclude": ["@babel/plugin-transform-regenerator"] }
        ],
        "@babel/react"
      ]
    }
  }
}
