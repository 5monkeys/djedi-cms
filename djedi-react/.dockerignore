*.log
*.md
*cache
.dockerignore
.git/
.gitignore
.next/
.stuff/
Dockerfile
dist/
docker-compose.yml
node_modules/
out/
static/
