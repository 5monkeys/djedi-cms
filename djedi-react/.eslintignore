!/*.js
!/**/.eslintrc.js
coverage
dist
*.snap
