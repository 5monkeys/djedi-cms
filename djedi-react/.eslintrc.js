const baseRules = require("eslint-config-lydell");

module.exports = {
  plugins: ["import", "react", "prettier", "sort-imports-es6-autofix"],
  parser: "babel-eslint",
  env: { es6: true },
  globals: {
    clearTimeout: false,
    console: false,
    document: false,
    fetch: false,
    setTimeout: false,
    window: false,
  },
  rules: Object.assign({}, baseRules({ import: true, react: true }), {
    "prettier/prettier": "error",
    "sort-imports-es6-autofix/sort-imports-es6": "error",
  }),
  overrides: [
    {
      files: [".*.js", "*.config.js"],
      env: { node: true },
    },
    {
      files: ["*.test.js", "{test,__mocks__}/*.js"],
      env: { node: true, jest: true },
    },
    {
      files: ["src/*.js"],
      rules: {
        "no-restricted-syntax": [
          "error",
          {
            selector:
              ":matches(FunctionDeclaration, FunctionExpression, ArrowFunctionExpression):matches([async=true], [generator=true])",
            message:
              "async functions and generators are not allowed, because it requires a big runtime in older browsers, which we donâ€™t want to force on all package consumers. Use `.then()` instead.",
          },
        ],
      },
    },
  ],
};
