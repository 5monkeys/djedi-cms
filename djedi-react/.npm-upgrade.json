{
  "ignore": {
    "prop-types": {
      "versions": "*",
      "reason": "No reason to bump this dependency"
    },
    "coffeescript": {
      "versions": "*",
      "reason": "Not compatible with newer versions"
    }
  }
}