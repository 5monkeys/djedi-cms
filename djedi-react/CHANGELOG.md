### Version 2.0.0 (2018-09-06)

- Fixed: Dedenting is now more intuitive. This is a breaking change. See below
  for more information.
- Fixed: Changes to `<Node>` defaults now update correctly during development.
  As a side effect, rendering the same node URI twice with different defaults no
  longer logs a warning (instead, the newer default replaces the old one).
- Improved: Documentation. More examples, clarifications and notes on quirks.

#### Dedenting

Previously, refactoring this:

<!-- prettier-ignore -->
```js
const node = (
    <Node uri="uri">{md`
        text
    `}</Node>
);
```

… into this:

<!-- prettier-ignore -->
```js
const text = md`
    text
`;
console.log(text);
const node = (
    <Node uri="uri">{text}</Node>
);
```

… was surprising. The `text` variable included the whitespace (as seen with the
`console.log`), but `<Node>` dedented it away. That caused the node to be
rendered differently when it was fetched by `<Node>` versus when `text` was
passed to `djedi.prefetch({ extra: [{ uri: "uri", value: text }] })`. Four
spaces of indentation can cause a code block in markdown.

Version 2.0.0 removes the auto-dedenting in `<Node>`. Instead the `md` tag now
dedents. This makes the above case more intuitive. As a side effect, the `md`
tag now allows interpolation and no longer warns about that, but the Babel
plugin still errors in that case so it doesn’t really matter.

When writing plain JSX text the dedenting in `<Node>` wasn’t needed anyway, due
to Babel’s JSX whitespace rules:

<!-- prettier-ignore -->
```js
<Node uri="uri">
    This is fine!
</Node>
```

In all other cases it is better to use the `md` tag anyway.

The "dedent" package has also been swapped for the "dedent-js" package. The
latter handles JS escapes correctly.

### Version 1.0.0 (2018-08-16)

- No changes, just marking as stable.

### Version 0.1.2 (2018-08-15)

- Fixed: The admin sidebar is now properly updated as nodes come and go on the
  page.

### Version 0.1.1 (2018-08-13)

- Fixed: Nodes with null values returned from the server are now handled
  correctly. This happens when there’s neither a default value nor a database
  value. Previously this resulted in an error being rendered. Now an empty node
  is rendered, as expected.

### Version 0.1.0 (2018-07-03)

- Initial release.
