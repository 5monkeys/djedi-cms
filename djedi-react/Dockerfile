FROM node:10

WORKDIR /code

COPY package.json package-lock.json ./
RUN npm ci

COPY . .

RUN npm run build

ENV PATH /code/node_modules/.bin:$PATH
