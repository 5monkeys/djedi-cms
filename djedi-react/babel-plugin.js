const dedent = require("dedent-js");

const MODULE_NAME = "djedi-react";
const CLIENT_NAME = "djedi";
const CLIENT_METHOD_NAME = "reportPrefetchableNode";
const COMPONENT_NAME = "Node";
const URI_ATTR_NAME = "uri";
const TEMPLATE_TAG_NAME = "md";
const NODE_URI_KEY = "uri";
const NODE_VALUE_KEY = "value";
const URI_VAR_NAME = "djedi_uri";
const DEFAULT_VAR_NAME = "djedi_default";

module.exports = function djediBabelPlugin({ types: t }) {
  return {
    name: MODULE_NAME,
    pre() {
      this.clientId = undefined;
      this.last = undefined;
    },
    post() {
      this.clientId = undefined;
      this.last = undefined;
    },
    visitor: {
      JSXElement(path) {
        // Bail if the element isn't a node.
        if (path.node.openingElement.name.name !== COMPONENT_NAME) {
          return;
        }

        const uriAttr = getUriAttr(path, t);

        // Bail if the uri prop is missing.
        if (uriAttr == null) {
          return;
        }

        const uri = getUri(uriAttr.node.value, t);

        // Bail if the uri prop isn't a static string.
        if (uri == null) {
          return;
        }

        const child = maybeGetLoneChild(path, t);
        const defaultValue =
          child == null ? undefined : getDefaultValue(child, t);

        // Bail if there is a child but it isn't a static string.
        if (child != null && defaultValue == null) {
          return;
        }

        const program = path.scope.getProgramParent();

        // Add import for djedi (the first time).
        if (this.clientId == null) {
          this.clientId = program.generateUidIdentifier(CLIENT_NAME);
          [this.last] = program.path.unshiftContainer(
            "body",
            makeClientImport(this.clientId, t)
          );
        }

        // Move the uri into a variable.
        const uriId = path.scope.generateUidIdentifier(URI_VAR_NAME);
        program.push({ id: uriId, init: t.stringLiteral(uri) });
        uriAttr.get("value").replaceWith(t.jSXExpressionContainer(uriId));

        let defaultId = undefined;

        // Move the default value into a variable (if any).
        if (defaultValue != null) {
          defaultId = path.scope.generateUidIdentifier(DEFAULT_VAR_NAME);
          program.push({
            id: defaultId,
            init: t.stringLiteral(dedent(defaultValue)),
          });
          child.replaceWith(t.jSXExpressionContainer(defaultId));
          for (const otherChild of path.get("children")) {
            if (otherChild !== child) {
              otherChild.remove();
            }
          }
        }

        // Insert a `reportPrefetchableNode` call.
        [this.last] = this.last.insertAfter(
          makeReportCall(this.clientId, uriId, defaultId, t)
        );
      },
    },
  };
};

function getUriAttr(jsxElementPath, t) {
  const attrs = jsxElementPath.get("openingElement.attributes");
  let result = undefined;

  for (const attr of attrs) {
    if (t.isJSXIdentifier(attr.node.name, { name: URI_ATTR_NAME })) {
      if (result != null) {
        throw attr.buildCodeFrameError(
          `<${COMPONENT_NAME}> must only specify the \`${URI_ATTR_NAME}\` prop once.`
        );
      }
      result = attr;
    } else if (t.isJSXIdentifier(attr.node.name, { name: "children" })) {
      throw attr.buildCodeFrameError(
        `<${COMPONENT_NAME}> must not have \`children\` as a prop.`
      );
    }
  }

  return result;
}

function getUri(value, t) {
  if (t.isStringLiteral(value)) {
    return value.value;
  }

  if (t.isJSXExpressionContainer(value)) {
    if (t.isStringLiteral(value.expression)) {
      return value.expression.value;
    }

    if (
      t.isTemplateLiteral(value.expression) &&
      value.expression.expressions.length === 0 &&
      value.expression.quasis.length === 1
    ) {
      return value.expression.quasis[0].value.cooked;
    }

    return undefined;
  }

  return undefined;
}

function maybeGetLoneChild(jsxElementPath, t) {
  const children = jsxElementPath.get("children");
  let result = undefined;

  for (const child of children) {
    // There can be whitespace-only JSXText nodes around an
    // JSXExpressionContainer. Ignore those, just like Babel does.
    // Also ignore `{/* comments */}`.
    if (!(isEmptyJSXText(child.node, t) || isJSXComment(child.node, t))) {
      if (t.isJSXElement(child.node)) {
        throw child.buildCodeFrameError(
          `<${COMPONENT_NAME}> only takes a single string as children. Wrap the default value in the \`md\` template tag to include HTML.`
        );
      }

      if (result != null) {
        throw child.buildCodeFrameError(
          `<${COMPONENT_NAME}> only takes a single string as children. Did you mean to use \`[foo]\` instead of \`{foo}\`?`
        );
      }
      result = child;
    }
  }

  return result;
}

function isEmptyJSXText(node, t) {
  return t.isJSXText(node) && node.value.trim() === "";
}

function isJSXComment(node, t) {
  return (
    t.isJSXExpressionContainer(node) &&
    t.isJSXEmptyExpression(node.expression) &&
    node.expression.innerComments != null &&
    node.expression.innerComments.length > 0
  );
}

function getDefaultValue(valuePath, t) {
  if (t.isJSXText(valuePath.node)) {
    return valuePath.node.value;
  }

  if (t.isJSXExpressionContainer(valuePath.node)) {
    if (t.isStringLiteral(valuePath.node.expression)) {
      return valuePath.node.expression.value;
    }

    const templateLiteral = t.isTemplateLiteral(valuePath.node.expression)
      ? valuePath.get("expression")
      : t.isTaggedTemplateExpression(valuePath.node.expression) &&
        t.isIdentifier(valuePath.node.expression.tag, {
          name: TEMPLATE_TAG_NAME,
        })
      ? valuePath.get("expression.quasi")
      : undefined;

    if (templateLiteral != null) {
      const expressions = templateLiteral.get("expressions");
      if (expressions.length > 0) {
        throw expressions[0].buildCodeFrameError(
          `Using \`\${foo}\` in a <${COMPONENT_NAME}> default value is an anti-pattern: it won't work if the user edits the node. Did you mean to use \`{foo}\` (without the \`$\`) or \`[foo]\`?`
        );
      }
      return templateLiteral.node.quasis[0].value.cooked;
    }

    return undefined;
  }

  return undefined;
}

function makeClientImport(clientId, t) {
  return t.importDeclaration(
    [t.importSpecifier(clientId, t.identifier(CLIENT_NAME))],
    t.stringLiteral(MODULE_NAME)
  );
}

function makeReportCall(clientId, uriId, defaultId, t) {
  return t.expressionStatement(
    t.callExpression(
      t.memberExpression(clientId, t.identifier(CLIENT_METHOD_NAME)),
      [
        t.objectExpression([
          t.objectProperty(t.identifier(NODE_URI_KEY), uriId),
          t.objectProperty(
            t.identifier(NODE_VALUE_KEY),
            defaultId == null ? t.identifier("undefined") : defaultId
          ),
        ]),
      ]
    )
  );
}
