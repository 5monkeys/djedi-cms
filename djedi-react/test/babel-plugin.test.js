import * as babel from "@babel/core";
import jsx from "@babel/plugin-syntax-jsx";
import checkDuplicatedNodes from "babel-check-duplicated-nodes";
import dedent from "dedent-js";
import fs from "fs";

import babelPlugin from "../babel-plugin";

// Make snapshots easier to read.
// Before: `"console.log(\\"Hello!\\")"`
// After: `console.log("Hello!")`
expect.addSnapshotSerializer({
  print(value) {
    return value;
  },
  test(value) {
    return typeof value === "string";
  },
});

function transform(code) {
  return babel.transform(code, {
    plugins: [jsx, babelPlugin],
    highlightCode: false,
  }).code;
}

function transformToAST(code) {
  return babel.transform(code, {
    plugins: [jsx, babelPlugin],
    ast: true,
    code: false,
  }).ast;
}

test("it works", () => {
  // The fixture is pretty long, so it makes sense not using an inline snapshot.
  const code = fs.readFileSync(require.resolve("./fixtures/nodes.js"), "utf8");
  expect(transform(code)).toMatchSnapshot();
});

test("it does not affect files with no <Node>s", () => {
  const code = dedent`
    console.log('<Node uri="uri">value</Node>');
  `;
  expect(transform(code)).toMatchInlineSnapshot(
    `console.log('<Node uri="uri">value</Node>');`
  );
});

// Comments at the start and end are currently supported, but not comments
// _between_ two JSXText nodes.
test("comments", () => {
  const code = dedent`
    <Node uri="uri">
      {/* comment */}
      value
      {
        // comment
        // comment
      }
    </Node>
  `;
  expect(transform(code)).toMatchInlineSnapshot(`
var _djedi_uri = "uri",
    _djedi_default = "value";
import { djedi as _djedi } from "djedi-react";

_djedi.reportPrefetchableNode({
  uri: _djedi_uri,
  value: _djedi_default
});

<Node uri={_djedi_uri}>{_djedi_default}</Node>;
`);
});

test("ensure no duplicated AST nodes are found", () => {
  const code = dedent`
    <>
      <Node uri="uri1">value1</Node>
      <Node uri="uri2">value2</Node>
    </>
  `;
  expect(() => checkDuplicatedNodes(babel, transformToAST(code))).not.toThrow(
    Error
  );
});

describe("it throws helpful errors", () => {
  test("duplicate uri prop", () => {
    expect(() => transform('<Node uri="overwritten uri" uri />'))
      .toThrowErrorMatchingInlineSnapshot(`
undefined: <Node> must only specify the \`uri\` prop once.
> 1 | <Node uri="overwritten uri" uri />
    |                             ^
`);
  });

  test("children prop", () => {
    expect(() =>
      transform('<Node uri="uri" children="children prop not supported" />')
    ).toThrowErrorMatchingInlineSnapshot(`
undefined: <Node> must not have \`children\` as a prop.
> 1 | <Node uri="uri" children="children prop not supported" />
    |                 ^
`);
  });

  test("jsx interpolation", () => {
    expect(() => transform('<Node uri="uri">JSX interpolation {nope}</Node>'))
      .toThrowErrorMatchingInlineSnapshot(`
undefined: <Node> only takes a single string as children. Did you mean to use \`[foo]\` instead of \`{foo}\`?
> 1 | <Node uri="uri">JSX interpolation {nope}</Node>
    |                                   ^
`);
  });

  test("empty jsx interpolation", () => {
    expect(() => transform('<Node uri="uri">empty {}</Node>'))
      .toThrowErrorMatchingInlineSnapshot(`
undefined: <Node> only takes a single string as children. Did you mean to use \`[foo]\` instead of \`{foo}\`?
> 1 | <Node uri="uri">empty {}</Node>
    |                       ^
`);
  });

  test("template literal interpolation", () => {
    expect(() =>
      transform(
        // eslint-disable-next-line no-template-curly-in-string
        '<Node uri="uri">{`template literal with interpolation ${nope}`}</Node>'
      )
    ).toThrowErrorMatchingInlineSnapshot(`
undefined: Using \`\${foo}\` in a <Node> default value is an anti-pattern: it won't work if the user edits the node. Did you mean to use \`{foo}\` (without the \`$\`) or \`[foo]\`?
> 1 | <Node uri="uri">{\`template literal with interpolation \${nope}\`}</Node>
    |                                                         ^
`);
  });

  test("md template literal interpolation", () => {
    expect(() =>
      transform(
        // eslint-disable-next-line no-template-curly-in-string
        '<Node uri="uri">{md`template literal with interpolation ${nope}`}</Node>'
      )
    ).toThrowErrorMatchingInlineSnapshot(`
undefined: Using \`\${foo}\` in a <Node> default value is an anti-pattern: it won't work if the user edits the node. Did you mean to use \`{foo}\` (without the \`$\`) or \`[foo]\`?
> 1 | <Node uri="uri">{md\`template literal with interpolation \${nope}\`}</Node>
    |                                                           ^
`);
  });

  test("accidental JSX children", () => {
    expect(() => transform('<Node uri="uri">A <em>mistake</em></Node>'))
      .toThrowErrorMatchingInlineSnapshot(`
undefined: <Node> only takes a single string as children. Wrap the default value in the \`md\` template tag to include HTML.
> 1 | <Node uri="uri">A <em>mistake</em></Node>
    |                   ^
`);
  });

  test("single JSX child", () => {
    expect(() => transform('<Node uri="uri"><em>mistake</em></Node>'))
      .toThrowErrorMatchingInlineSnapshot(`
undefined: <Node> only takes a single string as children. Wrap the default value in the \`md\` template tag to include HTML.
> 1 | <Node uri="uri"><em>mistake</em></Node>
    |                 ^
`);
  });
});
