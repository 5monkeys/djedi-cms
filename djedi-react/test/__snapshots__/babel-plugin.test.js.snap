// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it works 1`] = `
var _djedi_uri = "intro",
    _djedi_default = "[project] is a cool CMS",
    _djedi_uri2 = "en-us@link.txt",
    _djedi_default2 = "[text]",
    _djedi_uri3 = "simple",
    _djedi_uri4 = "string literal",
    _djedi_uri5 = "single quotes \\"",
    _djedi_uri6 = "simple template literal",
    _djedi_uri7 = "simple",
    _djedi_uri8 = "ok 1",
    _djedi_default3 = "simple",
    _djedi_uri9 = "ok 2",
    _djedi_default4 = "string literal",
    _djedi_uri10 = "ok 3",
    _djedi_default5 = "single quotes \\"",
    _djedi_uri11 = "ok 4",
    _djedi_default6 = "simple template literal",
    _djedi_uri12 = "ok 5",
    _djedi_default7 = "# markdown\\n\\nexample",
    _djedi_uri13 = "ok 6",
    _djedi_default8 = "simple jsx text on its own line (parses differently)",
    _djedi_uri14 = "ok 7",
    _djedi_default9 = "simple template literal on its own line (parses differently)";
import { djedi as _djedi2 } from "djedi-react";

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri,
  value: _djedi_default
});

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri2,
  value: _djedi_default2
});

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri3,
  value: undefined
});

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri4,
  value: undefined
});

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri5,
  value: undefined
});

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri6,
  value: undefined
});

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri7,
  value: undefined
});

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri8,
  value: _djedi_default3
});

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri9,
  value: _djedi_default4
});

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri10,
  value: _djedi_default5
});

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri11,
  value: _djedi_default6
});

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri12,
  value: _djedi_default7
});

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri13,
  value: _djedi_default8
});

_djedi2.reportPrefetchableNode({
  uri: _djedi_uri14,
  value: _djedi_default9
});

/* eslint-disable no-undef, react/jsx-no-undef */
import { Node, djedi, md } from "djedi-react";
import React from "react";
djedi.options.baseUrl = "/api";
const _djedi = "djedi";
const intro = <Node uri={_djedi_uri} project={_djedi}>{_djedi_default}</Node>;
export default function Home() {
  const link = text => <a href={\`?page=\${text}\`}>
      <Node uri={_djedi_uri2} text={text}>{_djedi_default2}</Node>
    </a>;

  return <div>
      {intro}

      <h2>ok uri</h2>
      <Node uri={_djedi_uri3} />
      <Node uri={_djedi_uri4} />
      <Node uri={_djedi_uri5} />
      <Node uri={_djedi_uri6} />
      <Node uri={_djedi_uri7} {...variables} />

      <h2>nope uri</h2>
      <Node uri={"nope".toUpperCase()} />
      <Node uri={\`template literal with interpolation \${nope}\`} />
      <Node uri={nope\`tagged template literal\`} />
      <Node uri={nope} />
      <Node uri={nope}>default</Node>

      <h2>ok children</h2>
      <Node uri={_djedi_uri8}>{_djedi_default3}</Node>
      <Node uri={_djedi_uri9}>{_djedi_default4}</Node>
      <Node uri={_djedi_uri10}>{_djedi_default5}</Node>
      <Node uri={_djedi_uri11}>{_djedi_default6}</Node>
      <Node uri={_djedi_uri12}>{_djedi_default7}</Node>
      <Node uri={_djedi_uri13}>{_djedi_default8}</Node>
      <Node uri={_djedi_uri14}>{_djedi_default9}</Node>

      <h2>nope children</h2>
      <Node uri="nope 1">{nope\`non-md tagged template literal\`}</Node>
      <Node uri="nope 2">{nope}</Node>

      <h2>non-Nodes</h2>
      <node uri="simple" />
      <TreeNode uri="simple" />

      <div>
        {link("previous")}
        {link("next")}
      </div>
    </div>;
}
`;
